File: 23_设计类/1236. 网络爬虫.py
from typing import List, Protocol
from collections import deque


class HtmlParser(Protocol):
    def getUrls(self, url: str) -> List[str]:
        raise NotImplementedError('Must implement getUrls()')


# 请你实现一个网络爬虫，以实现爬取同 startUrl 拥有相同 域名标签 的全部链接。该爬虫得到的全部链接可以 任何顺序 返回结果。
class Solution:
    def crawl(self, startUrl: str, htmlParser: 'HtmlParser') -> List[str]:
        def get_host(url: str):
            i0 = url.find("//")
            i1 = url.find("/", i0 + 2)
            return url[:i1]

        target = get_host(startUrl)
        queue = deque(htmlParser.getUrls(startUrl))
        res = {startUrl}
        while queue:
            cur = queue.popleft()
            if cur not in res:
                if cur.startswith(target):
                    res.add(cur)
                    queue.extend(htmlParser.getUrls(cur))
        return list(res)

File: 22_专题/多线程/1236. 网络爬虫.go
// O(V + E)，其中 V 是节点数（URL 数量），E 是边数（链接数量）

package main

import (
	"strings"
)

type HtmlParser interface {
	GetUrls(url string) []string
}

func crawl(startUrl string, htmlParser HtmlParser) []string {
	visited := make(map[string]struct{})
	queue := []string{}
	res := []string{}

	targetHostName := getHostName(startUrl)

	push := func(url string) {
		visited[url] = struct{}{}
		queue = append(queue, url)
		res = append(res, url)
	}
	push(startUrl)

	for len(queue) > 0 {
		url := queue[0]
		queue = queue[1:]
		for _, nextUrl := range htmlParser.GetUrls(url) {
			if _, has := visited[nextUrl]; !has && getHostName(nextUrl) == targetHostName {
				push(nextUrl)
			}
		}
	}

	return res
}

func getHostName(url string) string {
	return strings.Split(url, "/")[2]
}
